From: Jiang Xin <jiangxin@moon.ossxp.com>
Subject: [PATCH] t/load_mm_with_mmx_file

When load mindmap, join <filename>.mm file with .<filename>.mmx.

Signed-off-by: Jiang Xin <jiangxin@moon.ossxp.com>

---
 freemind/build.xml                                 |    7 +-
 freemind/freemind.bat                              |    2 +-
 freemind/freemind.sh                               |    1 +
 freemind/freemind/main/Tools.java                  |  119 ++++++++++++++++++++
 .../modes/mindmapmode/MindMapMapModel.java         |   23 +++-
 .../modes/mindmapmode/freemind_join_mm_mmx.xslt    |   36 ++++++
 6 files changed, 183 insertions(+), 5 deletions(-)

diff --git a/freemind/build.xml b/freemind/build.xml
index 94d57f3..3a3d179 100644
--- a/freemind/build.xml
+++ b/freemind/build.xml
@@ -57,8 +57,9 @@
 	<property name="xsltLib2" location="lib/serializer.jar"/>
 	<property name="xsltLib3" location="lib/xml-apis.jar"/>
 	<property name="xsltLib4" location="lib/xercesImpl.jar"/>
+	<property name="saxonLib" location="lib/saxon.jar"/>
 
-	<property name="classpath" value="${jibxruntimelibs}:${jibxjar}:${formsLib}:${simplyHtml}:${spellLib}:${xsltLib}:${xsltLib2}:${xsltLib3}:${xsltLib4}"/>
+	<property name="classpath" value="${jibxruntimelibs}:${jibxjar}:${formsLib}:${simplyHtml}:${spellLib}:${xsltLib}:${xsltLib2}:${xsltLib3}:${xsltLib4}:${saxonLib}"/>
 	<path id="jars.path">
 	  <fileset dir="${src}">
 	    <include name="**/*.jar" />
@@ -257,6 +258,7 @@
 				<include name="xercesImpl.jar"/>
 				<include name="forms-1.0.5.jar"/>
 				<include name="bindings.jar"/>
+				<include name="saxon.jar"/>
 				<include name="jibx/jibx-run.jar"/>
 				<include name="jibx/xpp3.jar"/>
 				<include name="SimplyHTML/gnu-regexp-1.1.4.jar"/>
@@ -452,7 +454,7 @@
 			<attribute name="Built-Number" value="${build.number}" />
 			<attribute name="Created-By" value="Joerg Mueller, Daniel Polansky, Christian Foltin, Dimitry Polivaev, and others." />
 			<attribute name="Main-Class" value="freemind.main.FreeMindStarter" />
-			<attribute name="Class-Path" value="freemind.jar .. commons-lang-2.0.jar forms-1.0.5.jar jibx/jibx-run.jar jibx/xpp3.jar bindings.jar jortho.jar xalan.jar serializer.jar xml-apis.jar xercesImpl.jar SimplyHTML/gnu-regexp-1.1.4.jar SimplyHTML/SimplyHTML.jar"/>
+			<attribute name="Class-Path" value="freemind.jar .. commons-lang-2.0.jar forms-1.0.5.jar jibx/jibx-run.jar jibx/xpp3.jar bindings.jar jortho.jar xalan.jar saxon.jar serializer.jar xml-apis.jar xercesImpl.jar SimplyHTML/gnu-regexp-1.1.4.jar SimplyHTML/SimplyHTML.jar"/>
 		</manifest>
 		<jar jarfile="${FM.jar}"
 		     manifest="${temp.manifest.file}"
@@ -472,6 +474,7 @@
 				<include name="Resources*"/>
 				<include name="mindmap_menus.xml"/>
 				<include name="**/freemind_version_updater.xslt"/>
+				<include name="**/freemind_join_mm_mmx.xslt"/>
 				<include name="dictionar*"/>
 			</fileset>
 		</jar>
diff --git a/freemind/freemind.bat b/freemind/freemind.bat
index 02bf65f..c8aa0bb 100644
--- a/freemind/freemind.bat
+++ b/freemind/freemind.bat
@@ -1,2 +1,2 @@
 @echo off
-java -Xmx256M -Xss8M -cp lib\freemind.jar;lib\commons-lang-2.0.jar;lib\forms-1.0.5.jar;lib\jibx\jibx-run.jar;lib\jibx\xpp3.jar;lib\bindings.jar;lib\xalan.jar;lib\serializer.jar;lib\xml-apis.jar;lib\xercesImpl.jar;lib\jortho.jar freemind.main.FreeMindStarter
+java -Xmx256M -Xss8M -cp lib\freemind.jar;lib\commons-lang-2.0.jar;lib\forms-1.0.5.jar;lib\jibx\jibx-run.jar;lib\jibx\xpp3.jar;lib\bindings.jar;lib\xalan.jar;lib\serializer.jar;lib\xml-apis.jar;lib\xercesImpl.jar;lib\jortho.jar;lib\saxon.jar freemind.main.FreeMindStarter
diff --git a/freemind/freemind.sh b/freemind/freemind.sh
index b101067..67863f6 100644
--- a/freemind/freemind.sh
+++ b/freemind/freemind.sh
@@ -206,6 +206,7 @@ ${freedir}/lib/xalan.jar:\
 ${freedir}/lib/serializer.jar:\
 ${freedir}/lib/xml-apis.jar:\
 ${freedir}/lib/xercesImpl.jar:\
+${freedir}/lib/saxon.jar:\
 ${freedir}"
 if [ "${JAVA_TYPE}" = "sun" ]
 then
diff --git a/freemind/freemind/main/Tools.java b/freemind/freemind/main/Tools.java
index f44904a..7a6da71 100644
--- a/freemind/freemind/main/Tools.java
+++ b/freemind/freemind/main/Tools.java
@@ -104,6 +104,12 @@ import freemind.modes.MindMapNode;
 import freemind.modes.mindmapmode.MindMapController;
 import freemind.view.mindmapview.NodeView;
 
+//OSSXP.COM: classes for .mm and .mmx join.
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.transform.dom.DOMSource;
+import org.w3c.dom.Document;
+
 /**
  * @author foltin
  * 
@@ -1085,6 +1091,119 @@ public class Tools {
 		return new BufferedReader(pReader);
 	}
 
+	/*
+	 * OSSXP.COM: hacked FreeMind saved two seperate files, .mm and .mmx file.
+	 * Join them in runtime using XSLT TransformerFactory.
+	 * TODO: Improvement needed. the joining stage may very slow, so disable it.
+	 */
+	public static Reader getActualReader(File file, FreeMindMain frame) throws IOException {
+	    if (!Resources.getInstance().getBoolProperty("wh_save_extra_attrs_in_aux_file"))
+	    {
+	    	return getActualReaderXml(file);
+	    }
+	    // load .mmx file...
+	    String ext = Tools.getExtension(file.getName());
+	    String mmxFileName = "";
+
+	    // OSSXP.COM: can disable join .mm with .mmx here.
+	    // if(true) return getActualReaderXml(file);
+	    
+	    if(!ext.equals("mm")) 
+	    {
+	    	mmxFileName = "." + file.getName()+".mmx";
+	    }
+	    else 
+	    {
+	    	mmxFileName = "." + Tools.removeExtension(file.getName()) + ".mmx";
+	    }
+	    File mmxfile = new File(file.getParent(), mmxFileName);
+	    
+	    if (!mmxfile.exists())
+	    {
+	    	return getActualReaderXml(file);
+	    }
+
+	    URL updaterUrl = null;
+	    InputStream inputStream = null;
+	    Source xsltSource = null;
+	    StringWriter buffwriter = null;
+	    Result result = null;
+	    TransformerFactory tf = null;
+	    Transformer transformer = null;
+	    String mmxFileFullName = mmxfile.toURI().toString();
+	    try {
+	        // try to convert map with xslt:
+	    	updaterUrl = frame.getResource(
+	                "freemind/modes/mindmapmode/freemind_join_mm_mmx.xslt");
+	        if (updaterUrl == null) {
+	            throw new IllegalArgumentException(
+	                    "freemind_join_mm_mmx.xslt not found.");
+	        }
+	        inputStream = updaterUrl.openStream();
+	        xsltSource = new StreamSource(inputStream);
+	        // get output:
+	        buffwriter = new StringWriter();
+	        result = new StreamResult(buffwriter);
+	        /* OSSXP.COM: create an instance of TransformerFactory.
+	         * the default xslt engine (com.sun.org.apache.xalan.internal.xsltc.trax...) 
+	         * may not support 'key()' in freemind_join_mm_mmx.xslt. 
+	         * Use saxon implement. */
+	        System.setProperty("javax.xml.transform.TransformerFactory", 
+	        		"com.icl.saxon.TransformerFactoryImpl");
+	        tf = TransformerFactory.newInstance();
+	        transformer = tf.newTransformer(xsltSource);
+	        transformer.setParameter("mmx_file", mmxFileFullName);
+	        transformer.transform(new StreamSource(file), result);
+
+	    } catch (Exception ex) {
+	        ex.printStackTrace();
+	        // exception: we take the file itself:
+	        return getActualReaderXml(file);
+	    } finally {
+	        if (inputStream != null) {
+	            inputStream.close();
+	        }
+	        if (buffwriter != null) {
+	            buffwriter.close();
+	        }
+	        inputStream = null;
+	        xsltSource = null;
+	        updaterUrl = null;
+	        result = null;
+	        transformer = null;
+	        tf = null;
+	    }
+	    return new StringReader(buffwriter.getBuffer().toString());
+	}
+
+	/* 
+	 * OSSXP.COM: In this hacked version, .mm file is a true XML file.
+	 * load XML file using DOM.
+	 */
+	private static Reader getActualReaderXml(File file) throws IOException {
+	    try
+	    {
+	        TransformerFactory tf = TransformerFactory.newInstance();
+	        Transformer transformer = tf.newTransformer();
+	        DocumentBuilderFactory domFactory = DocumentBuilderFactory.newInstance();
+	        DocumentBuilder domBuilder = domFactory.newDocumentBuilder();
+
+	        Document doc = domBuilder.parse(file);
+	        Source src = new DOMSource(doc);
+	        StringWriter buffwriter = new StringWriter();
+	        StreamResult result = new StreamResult(buffwriter);
+	        transformer.transform(src, result);
+
+	        return new StringReader(buffwriter.toString());
+	    }
+	    catch(Exception exp)
+	    {
+	    	exp.printStackTrace();
+	    }
+
+	    return new BufferedReader(new FileReader(file));
+	}
+
 	/**
 	 * In case of trouble, the method returns null.
 	 * 
diff --git a/freemind/freemind/modes/mindmapmode/MindMapMapModel.java b/freemind/freemind/modes/mindmapmode/MindMapMapModel.java
index a20c717..f8cbaeb 100644
--- a/freemind/freemind/modes/mindmapmode/MindMapMapModel.java
+++ b/freemind/freemind/modes/mindmapmode/MindMapMapModel.java
@@ -446,6 +446,10 @@ public class MindMapMapModel extends MapAdapter {
 		public String toString() {
 			return mString;
 		}
+
+    	public File getFile() {
+    		return null;
+    	}
 	}
 
 	private static class FileReaderCreator implements ReaderCreator {
@@ -462,10 +466,15 @@ public class MindMapMapModel extends MapAdapter {
 		public String toString() {
 			return mFile.getName();
 		}
+
+		public File getFile() {
+			return mFile;
+		}
 	}
 
 	public interface ReaderCreator {
 		Reader createReader() throws FileNotFoundException;
+    	File   getFile();
 	}
 
 	MindMapNodeModel loadTree(ReaderCreator pReaderCreator)
@@ -482,6 +491,9 @@ public class MindMapMapModel extends MapAdapter {
 				versionInfoLength);
 		// the resulting file is accessed by the reader:
 		Reader reader = null;
+        // OSSXP.COM: We need filename when do Tool.getActualReader,
+        //            for we need join .mm file with .mmx file.
+        File file = pReaderCreator.getFile();
 		for (int i = 0; i < EXPECTED_START_STRINGS.length; i++) {
 			versionInfoLength = EXPECTED_START_STRINGS[i].length();
 			String mapStart = "";
@@ -490,7 +502,10 @@ public class MindMapMapModel extends MapAdapter {
 			}
 			if (mapStart.startsWith(EXPECTED_START_STRINGS[i])) {
 				// actual version:
-				reader = Tools.getActualReader(pReaderCreator.createReader());
+                if (file != null)
+                    reader = Tools.getActualReader(file, getFrame());
+                else
+                    reader = Tools.getActualReader(pReaderCreator.createReader());
 				break;
 			}
 		}
@@ -520,7 +535,11 @@ public class MindMapMapModel extends MapAdapter {
 				this.getModeController()
 						.getFrame()
 						.out("Error on conversion. Continue without conversion. Some elements may be lost!");
-				reader = Tools.getActualReader(pReaderCreator.createReader());
+
+				if (file != null)
+						reader = Tools.getActualReader(file, getFrame());
+				else
+						reader = Tools.getActualReader(pReaderCreator.createReader());
 			}
 		}
 		try {
diff --git a/freemind/freemind/modes/mindmapmode/freemind_join_mm_mmx.xslt b/freemind/freemind/modes/mindmapmode/freemind_join_mm_mmx.xslt
new file mode 100644
index 0000000..bee253b
--- /dev/null
+++ b/freemind/freemind/modes/mindmapmode/freemind_join_mm_mmx.xslt
@@ -0,0 +1,36 @@
+<xsl:stylesheet version="1.0"
+   xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
+<!-- Usage:
+   xsltproc -stringparam mmx_file mindmap.mmx <this_xslt> mindmap.mm
+-->
+   <xsl:output method="xml" version="1.0" encoding="utf-8"
+       indent="no" />
+
+   <xsl:param name="mmx_file" />
+   <xsl:variable name="indexfile" select="document($mmx_file)" />
+
+   <xsl:key name="node-by-id" match="node" use="@ID"/>
+
+   <xsl:template match="map">
+       <map>
+           <xsl:copy-of select="@*" />
+           <xsl:apply-templates />
+       </map>
+   </xsl:template>
+
+   <xsl:template match="node">
+       <xsl:variable name="id" select="@ID" />
+       <xsl:copy>
+           <xsl:copy-of select="@*" />
+           <xsl:for-each select="$indexfile">
+               <xsl:copy-of select="key('node-by-id', $id)/@*" />
+           </xsl:for-each>
+           <xsl:apply-templates />
+       </xsl:copy>
+   </xsl:template>
+
+   <xsl:template match="*">
+     <xsl:copy-of select="."/>
+   </xsl:template>
+
+</xsl:stylesheet>
-- 
tg: (8cfa48e..) t/load_mm_with_mmx_file (depends on: tgmaster)
